internal/validator/validator.go

```diff
@@ -0,0 +1,302 @@
+
+// Package validator provides components to create epoch claims and update
+// rollups outputs with their proofs.
+package validator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"time"
+
+	"github.com/cartesi/rollups-node/internal/merkle"
+	. "github.com/cartesi/rollups-node/internal/node/model"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// The maximum height for the Merkle tree of all outputs produced
+// by an application
+const MAX_OUTPUT_TREE_HEIGHT = 63
+
+type ValidatorRepository interface {
+	// GetAllRunningApplications returns a slice with the applications currently
+	// being validated by the node.
+	GetAllRunningApplications(ctx context.Context) ([]Application, error)
+	// GetOutputsProducedInBlockRange returns outputs produced by inputs sent to
+	// the application in the provided block range, inclusive. Outputs are
+	// returned in ascending order by index.
+	GetOutputsProducedInBlockRange(
+		ctx context.Context,
+		application Address,
+		firstBlock, lastBlock uint64,
+	) ([]Output, error)
+	// GetProcessedEpochs returns epochs from the application which had all
+	// its inputs processed. Epochs are returned in ascending order by index.
+	GetProcessedEpochs(ctx context.Context, application Address) ([]Epoch, error)
+	// GetLastInputOutputHash returns the outputs Merkle tree hash calculated
+	// by the Cartesi Machine after it processed the last input in the epoch.
+	GetLastInputOutputHash(
+		ctx context.Context,
+		epochIndex uint64,
+		appAddress Address,
+	) (*Hash, error)
+	// GetPreviousEpoch returns the epoch that ended one block before the start
+	// of the current epoch
+	GetPreviousEpoch(ctx context.Context, currentEpoch Epoch) (*Epoch, error)
+	// ValidateEpochTransaction performs a database transaction
+	// containing two operations:
+	//
+	// 1. Updates an epoch, adding its claim and modifying its status.
+	//
+	// 2. Updates outputs with their Keccak256 hash and proof.
+	SetEpochClaimAndInsertProofsTransaction(
+		ctx context.Context,
+		epoch Epoch,
+		outputs []Output,
+	) error
+}
+
+// Validator creates epoch claims and rollups outputs proofs for all running
+// applications.
+type Validator struct {
+	repository              ValidatorRepository
+	inputBoxDeploymentBlock uint64
+}
+
+func NewValidator(
+	repo ValidatorRepository,
+	inputBoxDeploymentBlock uint64,
+) *Validator {
+	return &Validator{repo, inputBoxDeploymentBlock}
+}
+
+func (v *Validator) String() string {
+	return "validator"
+}
+
+// Run executes the Validator main logic of producing claims and/or proofs
+// for processed epochs of all running applications. It is meant to be executed
+// inside a loop. If an error occurs while processing any epoch, it halts and
+// returns the error.
+func (v *Validator) Run(ctx context.Context) error {
+	apps, err := v.repository.GetAllRunningApplications(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get running applications. %w", err)
+	}
+
+	for idx := range apps {
```

Comment:
Any strong reason to use `for idx := range apps` instead of `for _,app := range apps` ?

Reply:
To avoid creating an extra copy of the struct.